src/Build/Instance/TaskFactories/AssemblyTaskFactory.cs

```diff
@@ -424,6 +425,30 @@ internal ITask CreateTaskInstance(
             }
         }
 
+        /// <summary>
+        /// Overrides runtime/architecture with values from assembly metadata if available.
+        /// These values are more reliable since they are determined during assembly load.
+        /// </summary>
+        private TaskHostParameters UpdateTaskHostParameters(TaskHostParameters taskHostParameters)
+        {
+            // Determine runtime: prefer loaded type's runtime, fallback to parameter or current.
+            string runtime = !string.IsNullOrEmpty(_loadedType?.Runtime)
+                ? _loadedType.Runtime
+                : taskHostParameters.Runtime ?? XMakeAttributes.GetCurrentMSBuildRuntime();
+
+            // Determine architecture: prefer loaded type's architecture, fallback to parameter or current
```

Comment:
Not sure I understand the priority here. In my head the XML should always win--the user told us to do that thing, even if it doesn't make sense. Is that changing?